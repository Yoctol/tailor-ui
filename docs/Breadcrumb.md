---
id: breadcrumb
title: Breadcrumb
---

A breadcrumb displays the current location within a hierarchy. It allows going back to states higher up in the hierarchy.

## When To Use

- When the system has more than two layers in a hierarchy.
- When you need to inform the user of where they are.
- When the user may need to navigate back to a higher level.
- When the application has multi-layer architecture.

## Examples

```js
import { Breadcrumb } from 'tailor-ui';
```

### Basic

```jsx live
<Breadcrumb
  items={[
    {
      key: 'dashboard',
      name: 'Dashboard',
      onClick: () => {},
    },
  ]}
/>
```

### handle onClick

```jsx live
<Breadcrumb
  items={[
    {
      key: '1',
      name: 'Page 1',
      onClick: () => {},
    },
    {
      key: '2',
      name: 'Page 2',
      onClick: () => {},
    },
  ]}
/>
```

### Overflow

```jsx live
  <Breadcrumb
    items={[
      {
        key: '1',
        name: 'Page 1 is too too long long long long long long long long',
        onClick: () => {},
      },
      {
        key: '2',
        name: 'Page 2 is too too long long long long long long long long',
        onClick: () => {},
      },
      {
        key: '3',
        name: 'Page 3 is too too long long long long long long long long',
        onClick: () => {},
      },
      {
        key: '4',
        name: 'Page 4 is too too long long long long long long long long',
        onClick: () => {},
      },
    ]}
  />
```

## API

| Property | Description              | Type       | Default |
|----------|--------------------------|------------|---------|
| `items`  | Breadcrumb configuration | Breadcrumb | `[]`    |

### Breadcrumb item

| Property  | Description                             | Type                          | Default |
|-----------|-----------------------------------------|-------------------------------|---------|
| `key`     | key                                     | `string`                      |         |
| `name`    | display name                            | `ReactNode`                   |         |
| `onClick` | set the handler to handle `click` event | `(event: MouseEvent) => void` |         |
