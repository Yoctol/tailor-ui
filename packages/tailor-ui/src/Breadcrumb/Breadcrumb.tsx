import React, {
  FC,
  MouseEventHandler,
  ReactNode,
  createRef,
  forwardRef,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { MdKeyboardArrowRight, MdMoreHoriz } from 'react-icons/md';

import { useMeasure } from '@tailor-ui/hooks';

import { Box, Flex } from '../Layout';
import { Dropdown } from '../Dropdown';
import { Ellipsis } from '../Ellipsis';
import { Icon } from '../Icon';
import { Position } from '../constants';

import { MoreIcon, StyledBreadcrumbLink } from './styles';

export interface BreadcrumbLinkProps {
  active: boolean;
  name: ReactNode;
  lockWidth: boolean;
  onClick: MouseEventHandler<HTMLAnchorElement>;
}

const BreadcrumbLink = forwardRef<HTMLAnchorElement, BreadcrumbLinkProps>(
  function BreadcrumbLink({ active, name, lockWidth, onClick }, ref) {
    return (
      <StyledBreadcrumbLink
        ref={ref}
        active={active}
        lockWidth={lockWidth}
        onClick={onClick}
      >
        <Ellipsis>{name}</Ellipsis>
      </StyledBreadcrumbLink>
    );
  }
);

export interface BreadcrumbItem {
  key: string;
  name: ReactNode;
  onClick?: MouseEventHandler<HTMLAnchorElement>;
}

export interface BreadcrumbItemProps {
  items: BreadcrumbItem[];
}

const HiddenBreadcrumb: FC<BreadcrumbItemProps> = ({ items }) => {
  return (
    <Dropdown
      position={Position.BOTTOM}
      overlay={
        <Dropdown.List>
          {items.map(({ key, name, onClick }) => (
            <Dropdown.Item key={key} onClick={onClick}>
              {name}
            </Dropdown.Item>
          ))}
        </Dropdown.List>
      }
    >
      <MoreIcon type={MdMoreHoriz} fill="gray400" cursor="pointer" />
    </Dropdown>
  );
};

export interface BreadcrumbProps {
  items?: BreadcrumbItem[];
}

const Breadcrumb = forwardRef<HTMLDivElement, BreadcrumbProps>(
  function Breadcrumb({ items = [] }, forwardedRef) {
    const [{ ref }, bound] = useMeasure(forwardedRef);
    const [measured, setMeasured] = useState(false);
    const [hiddenKeys, setHiddenKeys] = useState<string[]>(
      items.map((item) => item.key)
    );

    const itemWithRefs = useMemo(
      () =>
        items.map((attrs) => ({
          ...attrs,
          itemRef: createRef<HTMLAnchorElement>(),
        })),
      [items]
    );

    useEffect(() => {
      setMeasured(false);

      if (
        bound.width !== 0 &&
        itemWithRefs.every(({ itemRef }) => itemRef.current?.offsetWidth)
      ) {
        let containerWidth = bound.width - 48;

        const keys = [...itemWithRefs]
          .reverse()
          .reduce<string[]>((prev, { key, itemRef }, index) => {
            const hasArrow = index !== 0;
            containerWidth -=
              (itemRef.current as HTMLAnchorElement).offsetWidth +
              (hasArrow ? 24 : 0);

            if (containerWidth < 0) {
              return [...prev, key];
            }

            return prev;
          }, []);

        setHiddenKeys(keys);
        setMeasured(true);
      }
    }, [bound.width, itemWithRefs]);

    const hiddenItems = items.filter(({ key }) => hiddenKeys.includes(key));

    if (items.length === 0) {
      return null;
    }

    return (
      <Flex
        ref={ref}
        flex="auto"
        alignItems="center"
        overflow="hidden"
        style={{
          opacity: measured ? 1 : 0,
          transition: 'transition: all 200ms ease 0s',
        }}
      >
        {hiddenItems.length > 0 && <HiddenBreadcrumb items={hiddenItems} />}
        {itemWithRefs.map(
          ({ key, itemRef, name, onClick = () => {} }, index) => {
            const hidden = hiddenKeys.includes(key);
            const active = index === itemWithRefs.length - 1;
            const arrow = index !== 0 && (
              <Icon type={MdKeyboardArrowRight} fill="gray400" />
            );

            return (
              <Box
                key={key}
                display="inline-flex"
                overflow="hidden"
                width={hidden ? 0 : 'auto'}
              >
                {arrow}
                <BreadcrumbLink
                  ref={itemRef}
                  active={active}
                  name={name}
                  lockWidth={items.length !== 1}
                  onClick={(event) => {
                    if (!active) {
                      onClick(event);
                    }
                  }}
                />
              </Box>
            );
          }
        )}
      </Flex>
    );
  }
);

export { Breadcrumb };
