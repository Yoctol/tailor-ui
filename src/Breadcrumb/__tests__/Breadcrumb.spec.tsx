import React, { createRef } from 'react';
import userEvent from '@testing-library/user-event';

import { render, screen } from 'test/test-utils';

import { Breadcrumb } from '../Breadcrumb';

describe('Breadcrumb', () => {
  // FIXME: the jsdom does not render real dom
  // so we can not get element width to render hidden menu
  it('should render breadcrumb correctly', () => {
    const { container } = render(
      <Breadcrumb
        items={[
          {
            key: '1',
            name: '1',
          },
        ]}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should trigger onClick correctly', async () => {
    const ref = createRef<HTMLDivElement>();
    const handleClick = jest.fn();

    render(
      <Breadcrumb
        ref={ref}
        items={[
          {
            key: '1',
            name: 'Click',
            onClick: handleClick,
          },
          {
            key: '2',
            name: 'Not clickable',
          },
        ]}
      />
    );

    const breadcrumb = await screen.findByText('Click');

    userEvent.click(breadcrumb);

    expect(handleClick).toBeCalled();
  });
});
