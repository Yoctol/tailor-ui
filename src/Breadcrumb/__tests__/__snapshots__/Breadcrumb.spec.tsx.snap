// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb should render breadcrumb correctly 1`] = `
.c3 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  width: 0%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: pointer;
}

.c1 svg {
  fill: #67758d;
  vertical-align: middle;
  pointer-events: none;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c1.c1.c1 svg {
  fill: #97a4ba;
  width: 24px;
  height: 24px;
}

.c5 {
  width: 100%;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 {
  max-width: auto;
  color: #97a4ba;
  font-size: 1rem;
  font-weight: bold;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  white-space: pre-wrap;
  cursor: auto;
  color: #283040;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover svg {
  fill: #328ef5 !important;
}

<div>
  <div
    class="c0"
    overflow="hidden"
    style="opacity: 0; transition: transition: all 200ms ease 0s;"
  >
    <i
      class="c1 c2"
      cursor="pointer"
      fill="gray400"
      size="24"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </i>
    <div
      class="c3"
      display="inline-flex"
      overflow="hidden"
      width="0"
    >
      <a
        class="c4"
      >
        <div
          class="c5"
        >
          1
        </div>
      </a>
    </div>
  </div>
</div>
`;
